# 05_Probability

### What is Probability?

In Artificial Intelligence and Machine Learning (AI/ML), probability is the mathematical framework for measuring how likely an event is to happen (scored from 0 to 1). Statistics is the mathematical science of gathering, analyzing, and drawing conclusions from data. Together, they form the core language that allows AI models to handle uncertainty, learn from noisy samples, and make predictions.

## 01. Introduction to Probability

**Definition**: Probability quantifies the likelihood of an event occurring. It is calculated as the ratio of favorable outcomes to total possible outcomes, assuming all outcomes are equally likely.

**Formula**: P(Event) = Number of Favorable Outcomes / Total Number of Possible Outcomes

**Real-World Example**: The probability of flipping a fair coin and getting heads is 1/2 = 0.5.

**Business Example**: An insurance company calculates the probability of a claim being filed based on historical data to set premium prices.

**AI/ML Example**: Classification models output probabilities (e.g., 85% chance an email is spam) rather than hard yes/no answers, allowing for confidence-based decisions.

In [2]:
favorable = 1
total = 2
p_heads = favorable / total
print("P(Heads):", p_heads)

P(Heads): 0.5


**Interpretation**: There's a 0.5 probability of getting heads — one favorable outcome out of two equally likely outcomes.

**Advantages**:
- Simple and intuitive foundation for all probability concepts.
- Directly applicable when outcomes are equally likely (dice, cards, coins).

**Limitations**:
- Classical probability assumes equally likely outcomes, which isn't always true in real life.
- Doesn't account for changing conditions or dependencies between events.

## 02. Events

**Definition**: An event is a specific outcome or set of outcomes from a random experiment. Events can be simple (single outcome, e.g., rolling a 4) or compound (multiple outcomes, e.g., rolling an even number).

**Formula**: P(Event) = Favorable Outcomes in Event / Total Outcomes in Sample Space

**Real-World Example**: Drawing a red card from a deck of 52 playing cards is an event with 26 favorable outcomes.

**Business Example**: "A customer completes a purchase after adding items to cart" is an event tracked by e-commerce companies for conversion rate analysis.

**AI/ML Example**: In classification tasks, each predicted class label (e.g., "spam" or "not spam") represents an event the model assigns a probability to.

In [3]:
sample_space = [1, 2, 3, 4, 5, 6]
event_even = [x for x in sample_space if x % 2 == 0]
p_event = len(event_even) / len(sample_space)
print("P(Even Number):", p_event)

P(Even Number): 0.5


**Interpretation**:
3 out of 6 outcomes (2, 4, 6) satisfy the "even number" event, giving a probability of 0.5.

**Advantages**:
- Provides a structured way to define and measure specific outcomes of interest.
- Forms the building block for more advanced probability rules (union, intersection, complement).

**Limitations**:
- Defining events clearly requires precise problem framing; ambiguous events lead to incorrect probability calculations.
- Compound events can become complex to enumerate as sample space grows.


## 03. Sample Space

**Definition**: The sample space is the set of all possible outcomes of a random experiment. It is denoted by S and forms the universe within which all events are defined.

**Formula**: S = {all possible outcomes}; P(S) = 1

**Real-World Example**: The sample space for rolling a six-sided die is S = {1, 2, 3, 4, 5, 6}.

**Business Example**: For a customer survey with responses "Satisfied," "Neutral," "Dissatisfied," the sample space is these three categories.

**AI/ML Example**: In multi-class classification, the sample space is the full set of possible class labels (e.g., all 10 digits in digit recognition, 0–9).

In [4]:
sample_space = [(i, j) for i in range(1, 7) for j in range(1, 7)]
print("Total Outcomes:", len(sample_space))

Total Outcomes: 36


**Interpretation**:
Rolling two dice produces 36 total possible outcomes — the full sample space within which all events are defined.

**Advantages**:
- Defines the complete boundary within which probability calculations are valid.
- Essential for correctly computing probabilities of complex or compound events.

**Limitations**:
- Can become very large and hard to enumerate for complex experiments (e.g., shuffling a deck of cards has 52! possible sample space arrangements).
- Assumes all elementary outcomes are well-defined and mutually exclusive.

## 04. Independent Events

**Definition**: Two events are independent if the occurrence of one does NOT affect the probability of the other occurring.

**Formula**: P(A and B) = P(A) × P(B)

**Real-World Example**: Flipping a coin twice — the result of the first flip has no effect on the second flip.

**Business Example**: A customer's decision to buy a product in Store A is independent of another unrelated customer's purchase decision in Store B.

**AI/ML Example**: Naive Bayes classifiers assume features are conditionally independent of each other given the class label — a simplifying assumption that makes computation efficient.

In [5]:
p_flip1 = 0.5
p_flip2 = 0.5
p_both = p_flip1 * p_flip2
print("P(Both Heads):", p_both)

P(Both Heads): 0.25


**Interpretation**:
The result of Flip 1 doesn't affect Flip 2, so the combined probability is simply the product: 0.5 × 0.5 = 0.25.

**Advantages**:
- Simplifies probability calculations significantly (simple multiplication rule).
- Widely applicable assumption in real-world sampling and experiments.

**Limitations**:
- Incorrectly assuming independence when events are actually dependent leads to inaccurate probability estimates.
- Rarely perfectly true in complex real-world systems with hidden correlations.

## 05. Dependent Events

**Definition**: Two events are dependent if the occurrence of one event affects the probability of the other occurring.

**Formula**: P(A and B) = P(A) × P(B | A)

**Real-World Example**: Drawing two cards from a deck without replacement — the probability of the second card depends on what the first card was.

**Business Example**: The probability a customer buys a warranty is dependent on whether they already purchased the associated electronic product.

**AI/ML Example**: In time-series forecasting (e.g., stock prices), today's value is dependent on previous values — models like ARIMA/LSTM explicitly account for this dependency.

In [6]:
p_first_ace = 4 / 52
p_second_ace = 3 / 51
p_both_aces = p_first_ace * p_second_ace
print("P(Both Aces):", round(p_both_aces, 4))

P(Both Aces): 0.0045


**Interpretation**:
The second draw's probability changed (3/51) because the first ace was removed from the deck — showing dependency between events.

**Advantages**:
- Accurately models real-world scenarios where events influence each other (sampling without replacement, sequential decisions).
- Forms the foundation for conditional probability and Bayesian reasoning.

**Limitations**:
- More complex to calculate than independent events; requires tracking how prior outcomes change subsequent probabilities.
- Errors compound if the dependency structure is misidentified.

## 06. Conditional Probability

**Definition**: Conditional probability is the probability of an event occurring given that another event has already occurred.

**Formula**: P(A | B) = P(A and B) / P(B)

**Real-World Example**: The probability it rains tomorrow given that it's cloudy today is higher than the unconditional probability of rain.

**Business Example**: The probability a customer churns given that they haven't logged in for 30 days is used by subscription companies to trigger retention campaigns.

**AI/ML Example**: Conditional probability underlies the entire framework of Bayesian models and is used to compute P(Class | Features) in classification tasks.

In [7]:
p_student_and_coffee = 0.30
p_student = 0.50
p_coffee_given_student = p_student_and_coffee / p_student
print("P(Coffee | Student):", p_coffee_given_student)

P(Coffee | Student): 0.6


**Interpretation**:
Given that a person is a student, there's a 60% chance they like coffee — higher than the overall average, since we're conditioning on known information.

**Advantages**:
- Enables more accurate predictions by incorporating relevant known information.
- Essential for real-world decision-making where context matters (medical diagnosis, targeted marketing).

**Limitations**:
- Requires sufficient data for the conditioning event to get a reliable estimate.
- Easy to misinterpret — confusing P(A|B) with P(B|A) is a common statistical error.

## 07. Bayes' Theorem

**Definition**: Bayes' Theorem describes how to update the probability of a hypothesis based on new evidence. It relates conditional probabilities in both directions.

**Formula**: P(A | B) = [P(B | A) × P(A)] / P(B)

Where:
- P(A) = Prior probability (before evidence)
- P(A|B) = Posterior probability (after evidence)
- P(B|A) = Likelihood
- P(B) = Evidence (total probability of B)

**Real-World Example**: Given a positive medical test result, Bayes' Theorem calculates the true probability of having the disease, accounting for test accuracy and disease rarity.

**Business Example**: Spam filters use Bayes' Theorem to calculate the probability an email is spam given the words it contains.

**AI/ML Example**: Naive Bayes Classifier directly applies Bayes' Theorem to predict class probabilities based on feature likelihoods.

In [8]:
p_disease = 0.01
p_pos_given_disease = 0.99
p_pos_given_no_disease = 0.05

p_positive = (p_pos_given_disease * p_disease) + (p_pos_given_no_disease * (1 - p_disease))
p_disease_given_positive = (p_pos_given_disease * p_disease) / p_positive
print("P(Disease | Positive Test):", round(p_disease_given_positive, 4))

P(Disease | Positive Test): 0.1667


**Interpretation**:
Even with a positive test, the true probability of having the disease is only ~16.7%, because the disease is rare — false positives from healthy people dilute the result.

**Advantages**:
- Provides a rigorous framework for updating predictions as new data arrives.
- Crucial for correctly interpreting diagnostic tests, spam filters, and probabilistic ML models.

**Limitations**:
- Requires an accurate prior probability, which can be difficult to estimate and heavily influences the result.
- Can be counter-intuitive and prone to misinterpretation (the "base rate fallacy").

## 08. Probability Rules

**Definition**: Probability rules are fundamental laws used to combine and calculate probabilities of multiple events, including the Addition Rule, Multiplication Rule, and Complement Rule.

**Formula**: 
- Addition Rule: P(A or B) = P(A) + P(B) − P(A and B)
- Multiplication Rule: P(A and B) = P(A) × P(B|A)
- Complement Rule: P(not A) = 1 − P(A)

**Real-World Example**: The probability of drawing a King OR a Heart from a deck uses the addition rule since some cards (King of Hearts) satisfy both conditions.

**Business Example**: The probability a marketing campaign succeeds via email OR social media reach uses the addition rule, accounting for overlap between channels.

**AI/ML Example**: Probability rules are used to combine outputs from ensemble models or compute joint probabilities in probabilistic graphical models.

In [9]:
p_king = 4 / 52
p_heart = 13 / 52
p_king_and_heart = 1 / 52
p_king_or_heart = p_king + p_heart - p_king_and_heart
print("P(King or Heart):", round(p_king_or_heart, 4))

P(King or Heart): 0.3077


**Interpretation**:
Adding P(King) and P(Heart) directly would double-count the King of Hearts, so it's subtracted once — giving a correct combined probability of ~0.3077.

**Advantages**:
- Provides systematic formulas to solve complex, real-world probability problems.
- Prevents common errors like double-counting overlapping outcomes.

**Limitations**:
- Applying the wrong rule (e.g., forgetting to subtract overlap) is a very common mistake.
- Requires correctly identifying whether events are mutually exclusive, independent, or overlapping before choosing a rule.

## 09. Random Variables

**Definition**: A random variable is a numerical value that represents the outcome of a random experiment. It can be **discrete** (countable values, e.g., number of heads in 3 flips) or **continuous** (any value in a range, e.g., height).

**Formula**: X: S → ℝ (maps each outcome in sample space S to a real number)

**Real-World Example**: The number of customers arriving at a store in an hour is a discrete random variable.

**Business Example**: Monthly revenue is a continuous random variable used in financial forecasting models.

**AI/ML Example**: Model outputs (e.g., predicted probability of default, predicted house price) are treated as random variables with associated probability distributions.

In [10]:
import random
flips = [random.choice([0, 1]) for _ in range(3)]
num_heads = sum(flips)
print("Random Variable X (Heads count):", num_heads)

Random Variable X (Heads count): 0


**Interpretation**:
X is a discrete random variable that maps the outcome of 3 coin flips to a number (0–3), representing how many heads occurred.

**Advantages**:
- Converts real-world random outcomes into numerical form suitable for mathematical/statistical analysis.
- Forms the foundation for probability distributions, expected value, and variance calculations.

**Limitations**:
- Discrete and continuous random variables require different mathematical treatment (PMF vs PDF), which can confuse beginners.
- Choosing the wrong distribution assumption for a random variable leads to inaccurate models.

## 10. Expected Value

**Definition**: Expected Value (EV) is the long-run average outcome of a random variable if an experiment is repeated many times. It is essentially a weighted average, where each outcome is weighted by its probability.

**Formula**: E(X) = Σ [x × P(x)]

**Real-World Example**: The expected value of a fair six-sided die roll is 3.5 — the long-run average if rolled many times.

**Business Example**: Insurance companies use expected value to calculate expected payout per policy and set profitable premium prices.

**AI/ML Example**: Expected value is used in reinforcement learning to calculate the expected future reward of taking an action in a given state.

In [11]:
outcomes = [1, 2, 3, 4, 5, 6]
probabilities = [1/6] * 6
expected_value = sum(x * p for x, p in zip(outcomes, probabilities))
print("Expected Value:", expected_value)

Expected Value: 3.5


**Interpretation**:
The expected value of 3.5 is the long-run average of rolling a fair die many times, even though 3.5 is not a possible single outcome.

**Advantages**:
- Provides a single, useful summary number for decision-making under uncertainty.
- Widely applicable in finance, gambling, insurance, and reinforcement learning for evaluating expected outcomes/rewards.

**Limitations**:
- Doesn't capture variability or risk — two scenarios can have the same expected value but very different risk profiles.
- Can be a misleading summary for highly skewed or rare-but-extreme outcome distributions.

## Interview Questions and Answers

**1. What is the difference between independent and dependent events?**
Independent events do not affect each other's probability — P(A and B) = P(A) × P(B). Dependent events do affect each other — the occurrence of one changes the probability of the other, calculated as P(A and B) = P(A) × P(B|A).

**2. Explain Bayes' Theorem in simple terms.**
Bayes' Theorem lets us update our belief about an event (posterior probability) after observing new evidence, by combining our initial belief (prior probability) with how likely the evidence is under that belief (likelihood).

**3. What is the difference between P(A|B) and P(B|A)? Why does confusing them matter?**
P(A|B) is the probability of A given B has occurred; P(B|A) is the reverse. Confusing them is called the "base rate fallacy" and can lead to serious errors — like overestimating disease probability from a positive test without considering how rare the disease is.

**4. What is a random variable, and what are its two types?**
A random variable is a numerical outcome of a random experiment. It can be discrete (countable values, like number of heads in flips) or continuous (any value in a range, like height or time).

**5. How is Expected Value calculated, and what does it represent?**
Expected Value is calculated as the sum of each outcome multiplied by its probability: E(X) = Σ[x × P(x)]. It represents the long-run average outcome if the experiment were repeated many times.

**6. What is the difference between the Addition Rule and Multiplication Rule in probability?**
The Addition Rule calculates P(A or B) and is used for combining probabilities of either event occurring. The Multiplication Rule calculates P(A and B) and is used for combining probabilities of both events occurring together.

**7. Why do Naive Bayes classifiers assume feature independence, and is this assumption realistic?**
Naive Bayes assumes independence to simplify probability calculations drastically, making the model fast and scalable. While this assumption is often unrealistic (features are frequently correlated), the model still performs well in practice for many tasks like text classification.

**8. What's the difference between sample space and an event?**
Sample space is the complete set of all possible outcomes of an experiment, while an event is a specific subset of outcomes we are interested in measuring the probability of.

## Self Reflection

- **Sample Space and Events** are the foundation everything else builds on — clearly defining them is the first (and most important) step.
- **Independent vs Dependent Events** require completely different formulas, and misidentifying which applies is a very common error.
- **Conditional Probability and Bayes' Theorem** were the most eye-opening — especially seeing how a "99% accurate" test can still give a low true probability when the underlying condition is rare.
- **Probability Rules** (addition, multiplication, complement) are simple individually, but combining them correctly for real-world problems takes practice.
- **Random Variables and Expected Value** bridge probability theory with practical decision-making, and are directly used in ML concepts like reinforcement learning rewards and model output probabilities.
